### DISCUSSION TITLE
How to get response type depending on the input params?

### DISCUSSION THREAD

**Discussion by pieter-berkel** [trpc/trpc#3945] [Q&A]
Simple example of a normal JS function with generics:
```
const example = <T extends unknown>(params: T): T => {
  return params
}

// TYPE: const params: readonly ["Hello", "World"]
const params = ['Hello', 'World']

// TYPE: const result: readonly ["Hello", "World"]
const result = example(params)

// TYPE: const str1: string, const str2: string
const [str1, str2] = result
```

This is what I want to achieve with TRPC. If I send some params I want my respone type be able to change depending on the params. How can I achieve that? 
```

// TYPE: const params: readonly ["Hello", "World"]
const params = ['Hello', 'World']

const { data } = trpc.router.model.example(params)

// I want the following type returned based on the params 
// TYPE: const result: readonly ["Hello", "World"]
const result = data;

// I want the same result as in the simple example
// TYPE: const str1: string, const str2: string
const [str1, str2] = result
```

I already checked this page in the docs, but this does not describe my use case (https://trpc.io/docs/infer-types)

**Comment by emdede** [comment]
Hi. I have a very similar issue. Somewhere along the way, .query seems to swallow some of the type information. My use-case is filtering an object based on an array of provided object keys.

I have created a simple example of a working implementation [here](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMoHsC2ECS5rxLIDeAUMsgORyUBcyAzmFKAOblUBGdjzbpAX1KkE6EE0ZYIAeU4ArCAjDIAvCWRx6lGOnSUANMk5bOcKJWRCRYiZgCeAMQCuIJcDGrkAHgDC6ADYAKnYADigQAB6QIAAmDMgA1hB26DBoUniQsIgQAHwAFGQUoCFOYPqC9HilYMFhDL4BdXkAlKq5JByi4sogEADusnKeJWUAdKL+TpjiY1AQMU5I+fmICIaTbSodRRTIQ4pgY3AMDMCsIKsI6+oA2pMAuvQMUgdK9wEPli0cFPNgTigIA01w4AkMxAEPz+EABQOQfUG8g08TwMGgACVYYCQF4MNhMgQcoY-EFQnlBMIwOTkGjMdigV5AshItE4sgsaIoDEvEwWCBWIY4CA7LlDABpZLxVkQWLxJIpNKBDpqXa3AAKUHQIWQoGQAFEolBEGAvArUshAhKpbknpaNVqQg9KaRqWFaSAas0GqTpVFZezzWl8bh8NkkCrOsVPWV6LsKDE4GBNBoRb9kJNpuJ6L7bs6KEIrN0JJAJGp7M5XGB3JdiLqY2A48hE8n6CAnP5-BsAlmGPRbtRKF8BN8qRAmMdkAB6KfIdAJV3jo6caez6BaqBAA)

Now with the exact same helper types and the exact same function body, the .query procedure does not narrow down my columns but always returns the full interface. For reference:
```ts
  test: publicProcedure
    .input(
      z.object({
        data: z.object({ fqdn: z.string() }),
        columns: z.array(Domain.keyof()),
      })
    )
    .query(
      async <ColType extends keyof Domain>({ input }: InputTypes<ColType>) => {
        const newObj = input.columns.reduce((acc, col) => {
          Object.assign(acc, { [col]: myDomain[col] })
          return acc
        }, {})
        return newObj as InferReturn<Domain, ColType>
      }
    ),
})

const test = await client.domain.test.query({
  data: { fqdn: 'example.com' },
  columns: ['fqdn'],
})
test.customer_id // No error ?!
```

Is there a way to make this work as expected?
